bundles/edu.kit.ipd.sdq.activextendannotations/src/edu/kit/ipd/sdq/activextendannotations/StaticDelegate.xtend

```diff
@@ -1,8 +1,382 @@
+							addWarning('''Type «simpleName» is listet multiple times.''')
+					]
+				}
+				val annotatedTypes = new HashSet(annotation.delegationTargets.map[type])
+				val duplicates = usedTypes.keySet.intersect(annotatedTypes)
+				if (duplicates.length == 0) {
+					validAnnotations += annotation
+				} else {
+					duplicates.forEach [ type |
+						(usedTypes.get(type) + #[annotation]).map[delegationTargetsExpression].forEach [
+							addError('''The type «type.simpleName» is also used in another @«StaticDelegate.simpleName» annotation.''')
+						]
+					]
+				}
+				annotatedTypes.forEach [ type |
+					usedTypes.computeIfPresent(type, [k, set|set += annotation; set]);
+					usedTypes.putIfAbsent(type, new HashSet(#[annotation]))
+				]
+			}
+			validAnnotations
+		}
+
+		def private checkForSelfReference(Iterable<? extends AnnotationReference> annotations,
+			MutableTypeDeclaration annotatedType) {
+			val validAnnotations = new ArrayList(annotations.length);
+			for (annotation : annotations) {
+				if (annotation.delegationTargets.map[type].contains(annotatedType)) {
+					annotation.getExpression('value').
+						addError('''A type cannot export its own methods. Remove the type «annotatedType.simpleName» from the list.''')
+				} else {
+					validAnnotations += annotation
+				}
+			}
+			validAnnotations
+		}
+
+		/**
+		 * Converts a {@link Visibility} instance to Xtend’s
+		 * {@link org.eclipse.xtend.lib.macro.declaration.Visibility}. Uses the
+		 * provided `defaultVisibility` if the instance is 
+		 * {@link Visibility.AS_DECLARED}.
+		 * 
+		 * @param visibility The visibility to convert.
+		 * @param defaultVisibility The visibility to use if `visibility` is
+		 * {@link Visibility.AS_DECLARED}
+		 */
+		def private static toXtendVisibility(Visibility visibility,
+			org.eclipse.xtend.lib.macro.declaration.Visibility defaultVisibility) {
+			switch (visibility) {
+				case PRIVATE:
+					org.eclipse.xtend.lib.macro.declaration.Visibility.PRIVATE
+				case PACKAGE: {
+					org.eclipse.xtend.lib.macro.declaration.Visibility.DEFAULT
+				}
+				case PUBLIC: {
+					org.eclipse.xtend.lib.macro.declaration.Visibility.PUBLIC
+				}
+				case AS_DECLARED: {
+					defaultVisibility
+				}
+			}
+		}
+
+		/**
+		 * Finds the duplicate entries in the provided `iterable`.
+		 * 
+		 * @param iterable The iterable to inspect
+		 * @return The set of elements present at least twice in the provided
+		 * `iterable`. Equality is checked based on {@link Object#equals}.
+		 */
+		def private static <T> duplicates(Iterable<T> iterable) {
+			val seen = new HashSet<T>(iterable.length)
+			val duplicates = new HashSet<T>
+			for (value : iterable) {
+				if (seen.contains(value)) {
+					duplicates += value
+				} else {
+					seen += value
+				}
+			}
+			duplicates
+		}
+
+		/**
+		 * @return The instersection of `a` and `b` in a new set.
+		 */
+		def private static <T> intersect(Set<T> a, Set<T> b) {
```

Comment:
probably overkill because it would add a dependency to XOCL but just to let you know:
see [edu.kit.ipd.sdq.xocl.extensions.SetXOCLExtensions.intersection(Set, Set)](https://github.com/kit-sdq/XOCL/blob/master/bundles/edu.kit.ipd.sdq.xocl.extensions/src/edu/kit/ipd/sdq/xocl/extensions/SetXOCLExtensions.xtend#L32)